cpp/ql/src/Security/CWE/CWE-014/MemsetMayBeDeleted.ql

```diff
@@ -0,0 +1,50 @@
+/**
+ * @name Call to `memset` may be deleted
+ * @description Using <code>memset</code> the function to clear private data in a variable that has no subsequent use
+ *              is potentially dangerous because the compiler can remove the call.
+ * @kind problem
+ * @id cpp/memset-may-be-deleted
+ * @problem.severity warning
+ * @precision high
+ * @tags security
+ *       external/cwe/cwe-14
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.EscapesTree
+import semmle.code.cpp.commons.Exclusions
+
+class MemsetFunction extends Function {
+  MemsetFunction() {
+    this.hasGlobalOrStdOrBslName("memset")
+    or
+    this.hasGlobalOrStdName("wmemset")
+    or
+    this.hasGlobalName(["bzero", "__builtin_memset"])
+  }
+}
+
+from FunctionCall call, LocalVariable v, MemsetFunction memset
+where
+  call.getTarget() = memset and
+  not isFromMacroDefinition(call) and
+  // `v` only escapes as the argument to `memset`.
+  forall(Expr escape | variableAddressEscapesTree(v.getAnAccess(), escape) |
+    call.getArgument(0) = escape.getUnconverted()
+  ) and
```

Comment:
I think this requirement is too strong, and it's likely the reason why CPP-Differences only shows `memset` calls to variables that aren't used at all. Most buffers will be arguments to `memcpy` or `printf` or something like that, and their address will escape there.

A simple solution could be to assume that every call to functions other than `memset` will read from `v` if it escapes. This can be improved by using our models library.

Reply:
> I think this requirement is too strong, and it's likely the reason why CPP-Differences only shows memset calls to variables that aren't used at all. Most buffers will be arguments to memcpy or printf or something like that, and their address will escape there.

That's a good point. I realized that this was a very strict requirement, but I hadn't made the connection to this affecting the results on CPP-differences.

> A simple solution could be to assume that every call to functions other than memset will read from v if it escapes. This can be improved by using our models library.

Yeah, I actually considered this approach. But I figured that it would be better to just teach `variableAddressEscapesTree` about the `AliasFunction` model after getting this query merged. But I see that there's not much value in the query without this additional model information.

Here's an LGTM run that compares the query in the PR with a version that uses the solution you proposed (at least I _think_ it is) with support from the models library: https://lgtm.com/query/2490354227721723019/ It produces actual good (i.e., sensitive-data related memsets) results on 3 projects.

Reply:
> But I figured that it would be better to just teach `variableAddressEscapesTree` about the `AliasFunction` model after getting this query merged.

That eventual enhancement should probably go in a new `variableAddressEscapes` predicate (without the `Tree`) since some callers of the original `variableAddressEscapesTree` don't use it as an escape analysis but are concerned precisely with the question of whether the variable address ends up at the top of some expression tree. An example of such a caller is the one I proposed in https://github.com/github/codeql/pull/5256#discussion_r582864306.